File: intersectionList.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    """
    TC: O(sum of length of both the lists) = O(n)
    SC: O(1)
    """
    def lenOfList(self, curr):
        size = 0
        
        while curr != None:
            size += 1
            curr = curr.next
        
        return size
        
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> ListNode:
        
        # 1. base case
        if headA == None and headB == None:
            return None
        
        # 2. when both the lists are not empty
        
        # calculate the length of both the lists
        sizeA = self.lenOfList(headA)
        sizeB = self.lenOfList(headB)
        
        # move the head of the largest list
        currA = headA
        currB = headB
        
        if sizeA > sizeB:
            diff = sizeA - sizeB
            
            while diff != 0:
                currA = currA.next
                diff -= 1
        elif sizeA < sizeB:
            diff = sizeB - sizeA
            
            while diff != 0:
                currB = currB.next
                diff -= 1
            
        # match the node's address to find the intersection
        
        while currA != None and currB != None:
            
            if currA == currB:
                return currA
            
            currA = currA.next
            currB = currB.next
        
        return None

File: intersection_list.java
//Time - O(n)
//Space - O(1)
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        ListNode curr = headA;
        ListNode curr1 = headB;
        int length=0;
        int length1=0;
        while(curr!=null){
            length++;
            curr=curr.next;
        }
        while(curr1!=null){
            length1++;
            curr1=curr1.next;
        }
        curr = headA;
        curr1 = headB;
        if(length>length1){
            int count = length-length1;
            while(count>0){
                curr=curr.next;
                count--;
            }
        }
        if(length1>length){
            int count = length1-length;
            while(count>0){
                curr1=curr1.next;
                count--;
            }
        }
        while(curr!=null && curr1!=null){
            if(curr==curr1){
                return curr;
            }
            curr=curr.next;
            curr1 = curr1.next;
        }
        return null;
    }
}


